### PR TITLE
feat(engine): Added `OptionFunc` and `With`

### PR DESCRIPTION
**Pull request by flc1125**
**I would like to add an `OptionFunc` to facilitate Gin configuration.**

> Note: This adjustment supports downward compatibility.

**Before:**

```go
package main

import "github.com/gin-gonic/gin"

func main() {
	r := gin.New()

	r.GET("/ping", func(c *gin.Context) {
		c.JSON(200, gin.H{
			"message": "pong",
		})
	})

	r.Run(":8002")
}
```

**But now:**

- New Engine

```go
package main

import "github.com/gin-gonic/gin"

func main() {
	var fc = func(e *gin.Engine) {
		e.GET("/ping", func(c *gin.Context) {
			c.JSON(200, gin.H{
				"message": "pong",
			})
		})
		
		e.Use(gin.Recovery())
	}

	r := gin.New(fc)

	r.Run(":8002")
}
```

- With Engine:

```go
package main

import "github.com/gin-gonic/gin"

func main() {
	r := gin.New()
	
	r.With(func(e *gin.Engine) {
		e.GET("/ping", func(c *gin.Context) {
			c.JSON(200, gin.H{
				"message": "pong",
			})
		})

		e.Use(gin.Recovery())
	})

	r.Run(":8002")
}
```

### CODE REVIEW

**Comment by flc1125** [comment]
@appleboy Does lint information, unrelated to PR, need to be processed?

**Comment by flc1125** [comment]
lint see #3584

**Comment by appleboy** [comment]
already fixed in #3583

**Comment by flc1125** [comment]
> already fixed in #3583, Please rebase the PR.

It's already done

**Comment by flc1125** [comment]
😢

**Comment by flc1125** [comment]
Okay, I'm tweaking related errors, this is devastating. 😂

**Comment by flc1125** [comment]
@appleboy Please try again. 3Q

**Review by appleboy — APPROVED** [review]

**Comment by appleboy** [comment]
@flc1125 Thanks.

File: gin.go [hunk 1]
```diff
@@ -311,6 +314,15 @@ func (engine *Engine) Use(middleware ...HandlerFunc) IRoutes {
 	return engine
 }
 
+// With returns a new Engine instance with the provided options.
```
Comment (moolex) [inline gin.go:317]: It's not new instance, just update current configuration

**Pull request merged**